torch/_dynamo/variables/inline_helper.py

```diff
@@ -0,0 +1,159 @@
+    for node in tx.output.graph.nodes.__reversed__():
+        num_nodes_need_update_metadata -= 1
+        if num_nodes_need_update_metadata < 0:
+            break
+        # restore the source_fn_stack to be nn module.
+        if "source_fn_stack" in node.meta and len(node.meta["source_fn_stack"]) > 0:
+            # below logic to get a unique name for source_fn_stack is mimic from
+            # the _Namespace.create_name() which is used to get a unique name for
+            # the fx node.
+            if tx.output.current_tracer not in tracer_to_used_names.keys():
+                # TODO(JackCaoG): use weakref here?
+                tracer_to_used_names[tx.output.current_tracer] = {}
+
+            base_module_key = module_vt.module_key.lower()
+
+            if (
+                base_module_key
+                not in tracer_to_used_names[tx.output.current_tracer].keys()
+            ):
+                tracer_to_used_names[tx.output.current_tracer][base_module_key] = 0
+
+            count = tracer_to_used_names[tx.output.current_tracer][base_module_key]
+            tracer_to_used_names[tx.output.current_tracer][base_module_key] += 1
+            unique_module_key = (
+                base_module_key if count == 0 else f"{base_module_key}_{count}"
+            )
+            node.meta["source_fn_stack"][-1] = (
+                unique_module_key,
+                type(module_vt.module),
+            )
+        # remove the additional stack trace caused by fwd inlining
+        if "stack_trace" in node.meta and len(node.meta["stack_trace"]) > 0:
+            splited = node.meta["stack_trace"].split("\n")
+            # handle the cases where make_fx is called.
+            if len(splited) > 7 and "_dynamo/variables/inline_helper.py" in splited[-5]:
+                node.meta["stack_trace"] = "\n".join(splited[:-7]) + "\n"
+            # handle the case for lifted parameters.
+            elif (
+                len(splited) > 4
+                and "return forward_call(*args, **kwargs)" in splited[-2]
+            ):
+                node.meta["stack_trace"] = "\n".join(splited[:-3]) + "\n"
+
+
+code_to_Fx = {}
+
+
+def decompose_and_inline_function_with_makefx(tx, fn, args, kwargs):
+    from functorch import make_fx
+
+    from torch._dispatch.python import enable_python_dispatcher
+    from torch._guards import detect_fake_mode
+    from .base import MutableLocal
+    from .builder import SourcelessBuilder
+    from .dicts import ConstDictVariable
+    from .lists import BaseListVariable
+
+    # convert arguments from VariableTracker to fake tensors + constants again
+    fake_value_args = []
+    for arg in args:
+        fake_value_args.append(vt_to_fake_helper(arg, tx))
+
+    fake_value_kwargs = {}
+    for key, value in kwargs.items():
+        fake_value_kwargs[key] = vt_to_fake_helper(value, tx)
+
+    # Wrap the function before calling make_fx to avoid make_fx modify the kwargs's key.
```

Comment:
huh

Reply:
oh yea, for some reason `make_fx` will append some index to the key of the `kwarg` which gave me a bunch of headache. I ended up create another wrapper to work around this issue.

Reply:
Do you have a code pointer for this?

Reply:
more details in https://github.com/pytorch/pytorch/pull/112296#issuecomment-1901432939

Reply:
and you actially suggested me to wrap the fx in https://github.com/pytorch/pytorch/pull/112296#issuecomment-1906240613 haha

Reply:
Oh, I think my explanation in my comment is clearer. The goal here is not "modify the kwargs key", it is "ensure that make_fx only takes in arguments positionally, so that there is a consistent calling convention that Dynamo can use"